hbase_sql/src/main/java/com/splicemachine/olap/OlapServerMaster.java

```diff
@@ -344,23 +324,68 @@ private void submitSparkApplication(Configuration conf) throws IOException, Inte
             ugi.checkTGTAndReloginFromKeytab();
             if (yarnUgi != null)
                 yarnUgi.checkTGTAndReloginFromKeytab();
+            if (exitWhenDone) {
+                int[] jobs = sparkContext.statusTracker().getActiveJobIds();
+                if (jobs.length == 0) {
+                    SpliceLogUtils.info(LOG, "No active jobs are running. Exiting ...");
+                    end.set(true);
+                    rzk.getZooKeeper().delete(queueZkPath, -1);
+                }
+            }
+
         }
 
         LOG.info("OlapServerMaster shutting down");
     }
 
-    private void publishServer(RecoverableZooKeeper rzk, String hostname, int port) throws InterruptedException, KeeperException {
+    private void publishServer(RecoverableZooKeeper rzk, String hostname, int port) throws InterruptedException, KeeperException, IOException {
         try {
             HostAndPort hostAndPort = HostAndPort.fromParts(hostname, port);
             queueZkPath = rzk.getZooKeeper().create(queueZkPath, Bytes.toBytes(hostAndPort.toString()),
                     ZooDefs.Ids.OPEN_ACL_UNSAFE, CreateMode.EPHEMERAL_SEQUENTIAL);
             rzk.getData(queueZkPath, new QueueWatcher(), null);
+
+            String root = HConfiguration.getConfiguration().getSpliceRootPath();
+            String queueRoot = root + HBaseConfiguration.OLAP_SERVER_PATH + HBaseConfiguration.OLAP_SERVER_QUEUE_PATH;
+            SpliceLogUtils.info(LOG, "Add a watcher to %s", queueRoot);
+            ZkUtils.getChildren(queueRoot, new NewOlapServerWatcher(queueRoot));
+
         } catch (Exception e) {
             LOG.error("Couldn't register OlapServer due to unexpected exception", e);
             throw e;
         }
     }
 
+    /**
+     * A watcher to detect whether a new OLAP server was submitted. If so, the current running one should exit
+     * once it completes all work
+     */
```

Comment:
This implies two can be running at the same time, but it doesn't look like given they are using the same ZK path for publishing the connection information

Reply:
The new olap server overwrites znode in publishServer()